workflow/controller/operator.go

```diff
@@ -82,15 +81,21 @@ func (wfc *WorkflowController) operateWorkflow(wf *wfv1.Workflow) {
 			woc.markWorkflowFailed(fmt.Sprintf("invalid spec: %s", err.Error()))
 			return
 		}
+	} else {
+		err = woc.podReconciliation()
```

Comment:
An entry is made into the wfQueue after every pod update. And every entry in the wfQueue other than the first one when the workflow starts will query all pods in the workflow and reconcile their status. Isn't that too expensive?

Reply:
There are couple of things that make it not too expensive, or at least less expensive than the previous design:
1. entries to workqueues are deduped. so subsequent adds do not have any affect if the workflow key already exists in the workqueue. This is a huge improvement over the previous channel approach.
2. podReconciliation applies filters in the query to ignore both completed and pending pods as part of the workflow. This limit the working set for both the api query, and prevents us from reprocessing a pod that we don't care about anymore.
3. pod updates will only enqueue workflows upon a state change, so reconciliation should only happen when something changes. In the near future, I want to be even smarter about the state change to see if the change is even something we care about (e.g. we should not care if pod goes from Initializing to Running)
4. previously, pod updates were updating workflows one by one after every pod update. This was a very slow process, resulted in a huge number of chatter with the API server, and was causing a huge backlog in the controller processing channel/loop. By re-listing pods in podReconciliation upon pod updates, there is a significant reduction in the number of api queries and updates.

Reply:
I should add that there is a consequence to this approach that I am still trying to work through. Because we defer examination of pod state to podReconciliation, there is now a chance for pods to become deleted from underneath the controller. I am encountering this in the pod limits testing when the controller is spending 3+ minutes just submitting pods, and by the time that the controller is done submitting, some of the pods ran to completion and were even deleted by kubernetes.

This is a consequence of using SharedInformer over regular informers. From their docs:

"SharedInformer has a shared data cache and is capable of distributing notifications for changes to the cache to multiple listeners who registered via AddEventHandler. If you use this, there is one behavior change compared to a standard Informer. When you receive a notification, the cache will be AT LEAST as fresh as the notification, but it MAY be more fresh. "

Reply:
I see. Since each pod update is creating an entry in the workflow queue, are these other side-effects like higher memory consumption, etc. In extreme cases, this could cause the controller to crash.

This certainly seems better than the previous approach. In subsequent commits, we should consider:

* Having separate queues for submit and update (some rate limiting could be applied to workflow creation)
* As you mentioned, only put entries in the workflow queue when pods complete (success or failure)
* In the pod queue, can we store the actual status of the pod when the event happened? That way, would avoid having to go the API server when the pod entry is processed. Also, this will eliminate the problem of the pod having been gc'd when the controller processed it.

Reply:
> are these other side-effects like higher memory consumption, etc.

There will be no increase in memory because with workqueues, work items are simple strings and are deduped. A second Add() to the queue with an existing key will not be created if the key already exists. This effectively means that the maximum number of items in the workqueue will be less than or equal to the number of workflows in the system. This is contrary to the channel approach because channels have an entry for *every* event, and cannot be not deduped.

> In the pod queue, can we store the actual status of the pod when the event happened? That way, would avoid having to go the API server when the pod entry is processed. 

It is not possible to store additional metadata/state information in workqueue because work items are simple strings. I imagine it would also make it difficult to dedup. But one benefit of using the SharedIndexInformer, it means that we can take advantage of the local cache instead of the actual API server query.